### PR TITLE
Prevent repository access to users without repository READ permission on team level [SCI-10972]

### PR DESCRIPTION
**Pull request by artoscinote**
Jira ticket: [SCI-10972](https://scinote.atlassian.net/browse/SCI-10972)

### What was done
fixed repository access
global search fixes
activity fixes

[SCI-10972]: https://scinote.atlassian.net/browse/SCI-10972?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/models/protocol.rb [hunk 1]
```diff
@@ -254,11 +254,7 @@ def self.viewable_by_user(user, teams, options = {})
   end
 
   def self.viewable_by_user_my_module_protocols(user, teams)
```
Comment (okriuchykhin) [inline app/models/protocol.rb:256]: We don't go below tasks with permissions, so it is not possible to assume that such protocols have user assignments. 
And this method is for linked to tasks protocols

File: app/models/asset.rb [hunk 2]
```diff
@@ -63,7 +63,7 @@ def self.search(
     options = {}
   )
 
-    teams = options[:teams] || current_team || user.teams.select(:id)
+    teams = options[:teams] || [current_team] || user.teams
```
Comment (aignatov-bio) [inline app/models/asset.rb:66]: `options[:teams]` always return array?

File: app/models/repository_row.rb [hunk 3]
```diff
@@ -128,7 +128,10 @@ def self.search(user,
                   _current_team = nil,
                   options = {})
 
-    teams = options[:teams] || current_team || user.teams.select(:id)
+    teams = (options[:teams] || [current_team] || user.teams.select(:id)).filter do |team|
```
Comment (aignatov-bio) [inline app/models/repository_row.rb:131]: Same here

**Review by aignatov-bio — CHANGES_REQUESTED** [review]
Will we need any changes in addons?

File: app/models/asset.rb [hunk 4]
```diff
@@ -74,7 +74,7 @@ def self.search(
                              .pluck(:id)
 
     assets_in_inventories = Asset.joins(repository_cell: { repository_column: :repository })
-                                 .where(repositories: { team: teams })
+                                 .where(repositories: { team: teams.filter { |t| t.permission_granted?(user, RepositoryPermissions::READ) } })
```
Comment (okriuchykhin) [inline app/models/asset.rb:77]: Why we can't use a subquery here?

Comment (okriuchykhin) [reply app/models/asset.rb:77]: And actually shouldn't we simply check RepositoryPermissions::READ on repositories?

**Pull request closed without merging**